pkg/monitoring/event/decode.go

```diff
@@ -0,0 +1,91 @@
+package event
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"regexp"
+)
+
+var programInvocation = regexp.MustCompile(`^Program\s([a-zA-Z0-9]+)?\sinvoke\s\[\d\]$`)
+var programFinished = regexp.MustCompile(`^Program\s([a-zA-Z0-9]+)?\s(?:success|error)$`)
+var programLogEvent = regexp.MustCompile(`^Program\s(?:log|data):\s([+/0-9A-Za-z]+={0,2})?$`)
+
+func ExtractEvents(logs []string, programIDBase58 string) []string {
+	invocationStack := []string{}
+	output := []string{}
+	for _, log := range logs {
+		if matches := programInvocation.FindStringSubmatch(log); matches != nil {
+			invokedProgramID := matches[1]
+			invocationStack = append(invocationStack, invokedProgramID)
+		} else if matches := programFinished.FindStringSubmatch(log); matches != nil {
+			finishedProgramID := matches[1]
+			if invocationStack[len(invocationStack)-1] == finishedProgramID {
+				invocationStack = invocationStack[:len(invocationStack)-1]
+			}
+		} else if matches := programLogEvent.FindStringSubmatch(log); matches != nil {
+			currentProgramID := invocationStack[len(invocationStack)-1]
+			if programIDBase58 == currentProgramID {
+				output = append(output, matches[1])
+			}
+		}
+	}
+	return output
+}
```

Comment:
@aalu1418 thoughts on this part. The idea is that that logs come in as a trace of the execution stack, ie. this contract call that contract and that contract calls this other contract, etc. (see this tests https://github.com/smartcontractkit/chainlink-solana/pull/362/files/3b6baf66f3aaddfefc74c36c0cee1745f5bfedad#diff-e30bc1b5acc41eed18060e1a802e77e8c3f4525e006aaee5ec55263e4e391438R14-R23) What I'm doing here is, identifying the log lines that have events in them from the contract we care about and not the other contracts in the call stack.

Reply:
i think it makes sense, maybe something for extra validation would be to test some logs that are messed up (missing information, incorrect format, etc) just to verify it behaves properly when there's unexpected logs

Reply:
Thanks. Yeah, currently only testing the happy path. Will add some more tests.